### ISSUE TITLE
Checksum on non-compressed sstables

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylla#497]
Currently, we only do checksum for sstables that have compression enabled. We should also do checksum for non-compressed sstables using the CRC component.

**Comment by tzach** [comment]
@slivne @raphaelsc is this duplicate with #168?
Do Cassandra have such mechanism?

**Comment by slivne** [comment]
No this is not a duplicate

when you use non compressed sstables there is a checksum file - as far as I
know when reading we do not validate that the checksum matches the data in
the file (allowing to find an error). We only write the checksum file.

**Comment by tzach** [comment]
Thanks @slivne promoting to high, as this may affect data integrity

**Comment by raphaelsc** [comment]
@slivne: Exactly. @nyh once had arguments about it. @nyh: Could you please share your thoughts with us?

**Comment by nyh** [comment]
Indeed, Cassandra verifies the checksum of **compressed** blocks it reads before uncompress them. Cassandra actually has a "compression option" crc_check_chance which allows not checking this every time, but it defaults to 1.0 and in Scylla we just assume it is 1.0 so we always check the checksum (see FIXME in sstables/compress.cc).

Regarding **uncompressed** files, the situation is more complex. According to my understanding of https://issues.apache.org/jira/browse/CASSANDRA-9839 and https://issues.apache.org/jira/browse/CASSANDRA-9947 and my grep of the code, Cassandra does **not** check the checksums on uncompressed files when reading from them. One of the reasons why not to check these checksums on read is that in uncompressed files you can read exactly the data you need, and don't need to read a whole 64 KB block at all, so why read such blocks just for the purpose of verifying the checksum?

Another question is why these checksums exist at all, and even stranger, why does the whole-file checksum exist. Grepping the Cassandra code, it seems to me (but I didn't yet find evidence in discussion in the web) that those are only used for "nodetool verify" and when streaming (so as not to stream broken data to a remote node). But it appears it isn't used in normal reads.

So perhaps we need to make sure we also verify the CRC on streaming and nodetool-verify.

By the way, if I understand correctly, XFS, unlike ZFS (for example), does *not* guarantee file data integrity - we can read from a file and get corrupt data. So perhaps even if Cassandra doesn't check checksums on uncompressed reads, maybe we should add an option to do so, with some loss of performance.

Another small point to think about: there is little reason for both compressed and uncompressed formats to exist, except backward compatibility. Going forward, we should probably pick one - which has the right balance of performance and reliability, and recommend that.

**Comment by glommer** [comment]
Is this still an issue ?

**Comment by fee-mendes** [comment]
So bump.

**Comment by raphaelsc** [comment]
> So bump.

Volunteering yourself to work on this? :-) I'd love to review the patch.

**Comment by fee-mendes** [comment]
I volunteer to anything, yet no promises when it will be available given these days. :smiling_face_with_tear:

**Comment by DoronArazii** [comment]
@mykaul don't think it's feasible for 5.3 - shall I push to the backlog?

**Comment by mykaul** [comment]
I'm not sure what is missing from https://github.com/scylladb/scylladb/commit/ea3e8d869d5c7e078077a7878459389265bdb5fe

**Comment by DoronArazii** [comment]
@denesb ^^

**Comment by mykaul** [comment]
OK, I doubt it'll be completed for 5.3 (and I'm somewhat concerned about the performance impact, but then we do compress by default these days). If we get it by 5.4, great. If not, we'll put it (back?) into the backlog.

**Comment by denesb** [comment]
The performance impact is the main thing holding this back. Implementing the check itself is easy, it can be done in 1-2 days. For uncompressed sstables, the checksums are stored in a separate component (`.crc32`), and we have two bad choices here:
* read it into memory (bloats in-memory footprints)
* read it from disk (adds >=1 I/O per sstable read)

**Comment by denesb** [comment]
> What I dislike about such options is that the user never really knows what values to put, apart from 1 or 0. There's no good indication of why 0.3 is better than 0.7 - not how much performance improvement/degradation and how much 'risk' is reduced/increased by steps that are not on/off.

I agree, I don't see much value in that, beyond a binary on/off switch. If there is a suspicion of corruption, we can just run `scylla sstable validate-checksums` on all sstables. We don't need a probabilistic check for that because there is nothing probabilistic about sstables being corrupted, they are either corrupt or not.

**Comment by vladzcloudius** [comment]
> > What I dislike about such options is that the user never really knows what values to put, apart from 1 or 0. There's no good indication of why 0.3 is better than 0.7 - not how much performance improvement/degradation and how much 'risk' is reduced/increased by steps that are not on/off.
> 
> I agree, I don't see much value in that, beyond a binary on/off switch. If there is a suspicion of corruption, we can just run `scylla sstable validate-checksums` on all sstables. We don't need a probabilistic check for that because there is nothing probabilistic about sstables being corrupted, they are either corrupt or not.

This is not about performance - this is about data integrity check which must be always enabled by default in every (!!!) configuration.

And data integrity must be checked **always** before you serve the read data to a user. If you don't - you may be serving corrupted data.

Bottom line: having `scylla sstable validate-checksums` is not enough - the check must be in every read path. AFAIU this **is** the way when compression is enabled.

**Comment by denesb** [comment]
Just found out that `crc_check_chance` is part of the compression options. So there is currently not even a way to set it for tables which don't use compression.
We could do this via the syntax:
```
AND compression = {'sstable_compression': '', `crc_check_chance': 0.7}
```

**Comment by denesb** [comment]
Update:
```
cqlsh> create TABLE  ks.tbl_crc ( pk int PRIMARY KEY) WITH compression = {'sstable_compression': '', 'crc_check_chance': '0.5'};
```
This suceeds and the table is created without compression.
However:
```
cqlsh> select crc_check_chance from system_schema.tables where keyspace_name = 'ks' and table_name = 'tbl_crc';

 crc_check_chance
------------------
                1

(1 rows)
```
We seem to ignore all the compression params, when setting a null compressor.

**Comment by denesb** [comment]
Closing as dup https://github.com/scylladb/scylladb/issues/2431. Even if we implement checksum checking for non-compressed sstables, we will want to make it opt-in, so `crc_check_changes` seems to be the way to go.

**Comment by nyh** [comment]
> Closing as dup [#2431](https://github.com/scylladb/scylladb/issues/2431).

Are you sure? That other issue is crc_check_chance in materialized views not being saved in the schema... How is it related to whether or not crc_check_chance is used (not the abilitity to set it!) on uncompressed tables?

> Even if we implement checksum checking for non-compressed sstables, we will want to make it opt-in, so `crc_check_changes` seems to be the way to go.

This is a big mess... Cassandra moved in 3.0 the crc_check_chance parameter from the compression parameters (where Scylla still defines it, right?) to stand-alone. See also discussion in https://issues.apache.org/jira/browse/CASSANDRA-18872. It seems (but I'm not sure) that Cassandra still doesn't check CRC for non-compressed tables even though the crc_check_chance parameter is now not in the compression parameters. I don't know why.  And we can decide that if we didn't fix it in 10 years, we never will. But it still doesn't mean this issue should be marked a duplicate of a not-very-related issue.

**Comment by denesb** [comment]
https://github.com/scylladb/scylladb/issues/2431 is not specific to MV. This schema property value (coming from CREATE TABLE or ALTER TABLE) is never propagated to the schema tables. We have the appropriate field in `schema::raw` but it is hard-coded to 0. This is true for all tables. Of course, we are also missing the code to act on a non-0 value.
